### PR TITLE
Make frodo-lib throw when errors occur and test this with msw

### PR DESCRIPTION
**Pull request by aphillipohl**
Hello, this is the start of my work on making the library throw and adding the ability (eventually) to test the OAuth2ClientOps with pollyjs. I'm trying to figure out a means to test the exception being thrown i.e. a HTTP Error like 500 or cannot connect/timeout from the requests to Forgerock so that we can see this happen.

I worry that in the axios layer we are also trapping and logging errors rather than throwing and letting the app deal with this but writing some further tests will prove this.

I am also wondering if pollyjs can actually cause HTTP errors as there do not seem to be docs on this. It might be better for us to set up our own fake server for this in the tests and then we can make it do whatever we like. I'll look into this with the wider team today!

### CODE REVIEW

File: src/api/BaseApi.ts [hunk 1]
```diff
@@ -158,7 +158,9 @@ export function generateAmApi(resource, requestOverride = {}) {
 
   // enable curlirizer output in debug mode
   if (state.getCurlirize()) {
+    console.log('fuck');
```
Comment (ashleyfrieze) [inline src/api/BaseApi.ts:161]: We probably want to remove this line.

File: src/api/OAuth2ClientApi.ts [hunk 2]
```diff
@@ -65,7 +66,27 @@ export async function putOAuth2Client(id, data) {
...
+  const urlString = util.format(
+    oauth2ClientURLTemplate,
+    state.getHost(),
+    getCurrentRealmPath(),
+    id
+  );
+  return generateAmApi(getApiConfig()).delete(urlString, {
+    withCredentials: true,
+  });
+}
```
Comment (ashleyfrieze) [inline src/api/OAuth2ClientApi.ts:92]: Is this on topic?

File: src/api/BaseApi.ts (hunk 1 above)
Comment (aphillipohl) [reply src/api/BaseApi.ts]: done

File: src/api/OAuth2ClientApi.ts (hunk 2 above)
Comment (aphillipohl) [reply src/api/OAuth2ClientApi.ts:92]: Well, I thought it was necessary to get pollyjs to work in the same way as the other tests. But I'll delete to for now.

Comment (aphillipohl) [reply src/api/OAuth2ClientApi.ts]: Removed

File: src/ops/utils/ExportImportUtils.ts [hunk 3]
```diff
@@ -118,7 +118,8 @@ export function saveToFile(type, data, identifier, filename) {
   }
   fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
     if (err) {
-      return printMessage(`ERROR - can't save ${type} to file`, 'error');
+      printMessage(`ERROR - can't save ${type} to file`, 'error');
+      throw err;
```
Comment (aphillipohl) [inline src/ops/utils/ExportImportUtils.ts:122]: throwing from within a callback here is silently hidden but the fs module. What is the correct way to propagate errors (and test them) here?

Comment (aphillipohl) [reply src/ops/utils/ExportImportUtils.ts:122]: everyone seems to recommend using the promise based fs module that allows throwing these exceptions... but do we really want to make this change everywhere?

File: src/test/snapshots/ops/ThemeOps.test.mjs.snap
Comment (aphillipohl) [inline src/test/snapshots/ops/ThemeOps.test.mjs.snap:1]: Not quite sure where this file has come from...

File: src/ops/utils/ExportImportUtils.ts (hunk 3 above)
Comment (ashleyfrieze) [reply src/ops/utils/ExportImportUtils.ts:122]: Let's consider refactoring this function to be `async` and see what the blast radius would be.

Comment (aphillipohl) [reply src/ops/utils/ExportImportUtils.ts:122]: It's used a lot of places but also mostly in async places. Let's see.

File: src/test/snapshots/ops/ThemeOps.test.mjs.snap
Comment (adam-cyclones) [reply src/test/snapshots/ops/ThemeOps.test.mjs.snap:1]: https://github.com/facebook/jest/blob/main/examples/snapshot/__tests__/__snapshots__/link.test.js.snap

A snap file comes from a test tool called Jest, if you make changes your snapshots will update.

File: src/ops/utils/ExportImportUtils.ts (hunk 3 above)
Comment (aphillipohl) [reply src/ops/utils/ExportImportUtils.ts]: We actually no longer need to change this as we are not calling the export/backup anymore.

**Comment by vscheuber** [comment]
Hi, I wanted to make a quick comment that may impact your work on this PR:

An ongoing effort is to properly and completely split the CLI from LIB. The files you are working on currently, the OAuth2ClientOps.ts, is one of the remaining areas where that split hasn't been completed.

This manifests mainly in operations involving the file system. All the functions for importing from and exporting to files belong in the CLI. The LIB should only import and export JSON documents.

What that means for OAuth2ClientOps.ts is that all its current methods should move into a file of the same name but in the CLI project but all the API logic should be extracted into its own functions and remain in the LIB file:

## Move to CLI
- listOAuth2Clients
- exportOAuth2ClientToFile
- exportOAuth2ClientsToFile
- exportOAuth2ClientsToFiles
- importOAuth2ClientsFromFile

## Create in LIB
- creaateOAuth2ClientExportTemplate
- getOAuth2Clients
- getOAuth2Client
- putOAuth2Client
- deleteOAuth2Client
- exportOAuth2Clients
- exportOAuth2Client
- importOAuth2Clients
- importOAuth2Client

I didn't realize y'all would start with the OAuth2ClientOps.ts file and I honestly didn't remember it was one of the outstanding files.

What I will do, and I hope it does not interfere too much with your work, is I will make the changes mentioned above and you can rebase your PR.

If you want to look at IdpOps.ts as an example, that is close to what the final structure will be.

File: src/test/snapshots/ops/ThemeOps.test.mjs.snap
Comment (aphillipohl) [reply src/test/snapshots/ops/ThemeOps.test.mjs.snap:1]: Yes I understand that, however the fact that this file appears when I haven't changed something that I think should affect this is the issue! Thanks for the help!

**Comment by aphillipohl** [comment]
Hey @vscheuber I've finally got everything working as I wanted, which was much more difficult that I expected to get the tests working and not interfering with PollyJS. Anyway let me know what you think, we will need to work on Journeys next but that can be another PR. Let me know when you make your changes to rearrange the organisation of the OAuth2Clients.

**Comment by aphillipohl** [comment]
Hi @vscheuber do you want to get on a call today and discuss things?

**Pull request closed without merging**